mistralrs-core/src/vision_models/llava/llava15.rs

```diff
@@ -0,0 +1,333 @@
+        num_image_tokens: usize,
+    ) -> Result<Tensor> {
+        let image_indexes = input_ids
+            .squeeze(0)?
+            .lt(0i64)?
+            .nonzero()?
+            .squeeze(1)?
+            .to_vec1::<u32>()?;
+        let image_ids = image_indexes
+            .iter()
+            .map(|x| Ok(-(input_ids.i((0, *x as usize))?.to_scalar::<i64>()?)))
+            .collect::<Result<Vec<i64>>>()?;
+        let mut result = input_ids.clamp(0i64, i64::MAX)?.to_dtype(DType::U32)?;
+        result = self.llm.embed(&result)?; //[seq_len,hidden_size]
+        let image_features = self.encode_images(&images.to_dtype(self.dtype)?)?; //[num of images,patch_size*patch_size,hidden_size]
+        let num_of_images = image_features.shape().dims()[0];
+        let mut image_features_vec = Vec::new();
+        for i in 0..num_of_images {
+            image_features_vec.push(image_features.get(i)?.unsqueeze(0)?);
+        }
+        for (i, image_index) in image_indexes.iter().enumerate() {
+            let image_id = image_ids[i];
+            result = result.slice_assign(
+                &[
+                    &(0usize..1usize),
+                    &(*image_index as usize..*image_index as usize + num_image_tokens),
+                    &(..),
+                ],
+                &image_features_vec[(image_id - 1) as usize],
+            )?;
+        }
+        //truncate
+        let (_, seq_len) = input_ids.shape().dims2()?;
+        if seq_len > self.config.text_config.max_length {
+            result = result.i((.., ..self.config.text_config.max_length, ..))?
+        }
+        Ok(result)
+    }
+
+    pub fn forward_inputs(
+        &self,
+        input_ids: &Tensor,
+        pixel_values: Option<Tensor>,
+        num_image_tokens: Option<usize>,
+        seqlen_offsets: &[usize],
+        start_offsets_kernel: Tensor,
+        context_lens: Vec<(usize, usize)>,
+        position_ids: Vec<usize>,
+    ) -> Result<Tensor> {
+        if let Some(ref pixel_values) = pixel_values {
+            // we assume(as it should be) only prompt request contains image
+            let input_embeds = self.prepare_inputs_labels_for_multimodal(
+                input_ids,
+                pixel_values,
+                num_image_tokens.unwrap(),
+            )?;
+            self.llm.forward_input_embed(
+                input_ids,
+                input_embeds,
+                seqlen_offsets,
+                start_offsets_kernel,
+                context_lens,
+            )
+        } else {
+            self.llm.forward(
+                input_ids,
+                seqlen_offsets,
+                start_offsets_kernel,
+                context_lens,
+                position_ids,
+            )
+        }
+    }
+}
+
+impl IsqModel for Model {
+    fn get_tensors(&mut self) -> (Vec<(&mut QMatMul, Option<usize>)>, &dyn DeviceMapper) {
+        // I don't really get this part
```

Comment:
Basically, we are getting the MatMul tensors which can be quantized. When you do self.llm.get_tensors, you are only going to ISQ the text model, which should be fine.

Reply:
removed